### ISSUE TITLE
Using to_yaml breaks ansible-lint

### ISSUE THREAD

**Issue by abeluck** [ansible/ansible-lint#776]
##### Summary
Using ansible-lint on a playbook with `to_yaml` fails with the error "`yaml.representer.RepresenterError: ('cannot represent an object', AnsibleUndefined)`"
##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details

```console (paste below)
$ ansible --version
ansible 2.9.6
  config file = /home/user/ops/ansible/ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/.local/lib/python3.7/site-packages/ansible
  executable location = /home/user/.local/bin/ansible
  python version = 3.7.3 (default, Dec 20 2019, 18:57:59) [GCC 8.3.0]

$ ansible-lint --version
ansible-lint 4.2.0
```

- ansible installation method: pip
- ansible-lint installation method: pip

##### OS / ENVIRONMENT
Debian 10 buster

##### STEPS TO REPRODUCE
1. Run the ansible-lint on the sample playbook
`ansible-lint playbooks/to-yaml-bug.yml`

**`to-yaml-bug.yml`:**
```yaml
---
- name: to_yaml bug
  hosts: localhost
  tasks:
    - name: dump - this works with the to from json workaround)
      copy:
        content: |
          {{ vars | to_json | from_json | to_yaml }}
        dest: /tmp/ansible-vars.yml
      changed_when: false

    - name: dump - fails
      copy:
        content: |
          {{ vars | to_yaml }}
        dest: /tmp/ansible-vars.yml
      changed_when: false

```

##### Desired Behaviour
ansible-lint exits without error.

##### Actual Behaviour

ansible-lint crashes with:

```console
$ ansible-lint playbooks/to-yaml-bug.yml
Traceback (most recent call last):
  File "/home/user/.local/bin/ansible-lint", line 10, in <module>
    sys.exit(main())
  File "/home/user/.local/lib/python3.7/site-packages/ansiblelint/__main__.py", line 187, in main
    matches.extend(runner.run())
  File "/home/user/.local/lib/python3.7/site-packages/ansiblelint/__init__.py", line 267, in run
    for child in ansiblelint.utils.find_children(arg, self.playbook_dir):
  File "/home/user/.local/lib/python3.7/site-packages/ansiblelint/utils.py", line 163, in find_children
    for child in play_children(basedir, item, playbook[1], playbook_dir):
  File "/home/user/.local/lib/python3.7/site-packages/ansiblelint/utils.py", line 214, in play_children
    fail_on_undefined=False)
  File "/home/user/.local/lib/python3.7/site-packages/ansiblelint/utils.py", line 182, in template
    **dict(kwargs, fail_on_undefined=fail_on_undefined))
  File "/home/user/.local/lib/python3.7/site-packages/ansiblelint/utils.py", line 73, in ansible_template
    return templar.template(varname, **kwargs)
  File "/home/user/.local/lib/python3.7/site-packages/ansible/template/__init__.py", line 611, in template
    ) for v in variable]
  File "/home/user/.local/lib/python3.7/site-packages/ansible/template/__init__.py", line 611, in <listcomp>
    ) for v in variable]
  File "/home/user/.local/lib/python3.7/site-packages/ansible/template/__init__.py", line 623, in template
    disable_lookups=disable_lookups,
  File "/home/user/.local/lib/python3.7/site-packages/ansible/template/__init__.py", line 623, in template
    disable_lookups=disable_lookups,
  File "/home/user/.local/lib/python3.7/site-packages/ansible/template/__init__.py", line 578, in template
    disable_lookups=disable_lookups,
  File "/home/user/.local/lib/python3.7/site-packages/ansible/template/__init__.py", line 837, in do_template
    res = j2_concat(rf)
  File "<template>", line 12, in root
  File "/home/user/.local/lib/python3.7/site-packages/ansible/plugins/filter/core.py", line 66, in to_yaml
    transformed = yaml.dump(a, Dumper=AnsibleDumper, allow_unicode=True, default_flow_style=default_flow_style, **kw)
  File "/home/user/.local/lib/python3.7/site-packages/yaml/__init__.py", line 290, in dump
    return dump_all([data], stream, Dumper=Dumper, **kwds)
  File "/home/user/.local/lib/python3.7/site-packages/yaml/__init__.py", line 278, in dump_all
    dumper.represent(data)
  File "/home/user/.local/lib/python3.7/site-packages/yaml/representer.py", line 27, in represent
    node = self.represent_data(data)
  File "/home/user/.local/lib/python3.7/site-packages/yaml/representer.py", line 58, in represent_data
    node = self.yaml_representers[None](self, data)
  File "/home/user/.local/lib/python3.7/site-packages/yaml/representer.py", line 231, in represent_undefined
    raise RepresenterError("cannot represent an object", data)
yaml.representer.RepresenterError: ('cannot represent an object', AnsibleUndefined)
```
```console
$ ansible-playbook --syntax-check  playbooks/test2.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

playbook: playbooks/test2.yml
```

##### Related bugs:

* Closed Nov 1, 2019 https://github.com/ansible/ansible-lint/issues/562
* Closed Jan 19, 2017 https://github.com/ansible/ansible/issues/20253

**Comment by webknjaz** [comment]
Have you tried ansible-lint from the master branch?

**Comment by lucastheisen** [comment]
I'm getting the same problem, but I don't have `to_yaml`, so that may be a red-herring, or my case may be separate.  That said, the failure message is the same:

```text
$ ansible-lint deploy_environment.yml
Traceback (most recent call last):
  File "/home/ltheisen/ansible-python/bin/ansible-lint", line 8, in <module>
    sys.exit(main())
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansiblelint/__main__.py", line 187, in main
    matches.extend(runner.run())
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansiblelint/__init__.py", line 267, in run
    for child in ansiblelint.utils.find_children(arg, self.playbook_dir):
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansiblelint/utils.py", line 163, in find_children
    for child in play_children(basedir, item, playbook[1], playbook_dir):
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansiblelint/utils.py", line 214, in play_children
    fail_on_undefined=False)
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansiblelint/utils.py", line 182, in template
    **dict(kwargs, fail_on_undefined=fail_on_undefined))
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansiblelint/utils.py", line 73, in ansible_template
    return templar.template(varname, **kwargs)
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/template/__init__.py", line 615, in template
    ) for v in variable]
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/template/__init__.py", line 615, in <listcomp>
    ) for v in variable]
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/template/__init__.py", line 627, in template
    disable_lookups=disable_lookups,
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/template/__init__.py", line 627, in template
    disable_lookups=disable_lookups,
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/template/__init__.py", line 627, in template
    disable_lookups=disable_lookups,
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/template/__init__.py", line 582, in template
    disable_lookups=disable_lookups,
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/template/__init__.py", line 841, in do_template
    res = j2_concat(rf)
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/template/native_helpers.py", line 29, in ansible_native_concat
    head = list(islice(nodes, 2))
  File "<template>", line 14, in root
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible/plugins/filter/core.py", line 72, in to_nice_yaml
    transformed = yaml.dump(a, Dumper=AnsibleDumper, indent=indent, allow_unicode=True, default_flow_style=False, **kw)
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/yaml/__init__.py", line 290, in dump
    return dump_all([data], stream, Dumper=Dumper, **kwds)
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/yaml/__init__.py", line 278, in dump_all
    dumper.represent(data)
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/yaml/representer.py", line 27, in represent
    node = self.represent_data(data)
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/yaml/representer.py", line 58, in represent_data
    node = self.yaml_representers[None](self, data)
  File "/home/ltheisen/ansible-python/lib64/python3.6/site-packages/yaml/representer.py", line 231, in represent_undefined
    raise RepresenterError("cannot represent an object", data)
yaml.representer.RepresenterError: ('cannot represent an object', AnsibleUndefined)
```

My playbook:
```yaml
---
- hosts: localhost
  gather_facts: no
  connection: local

  tasks:
  - name: Ensure we have a coord
    fail:
      msg: "Need a coord to use! Supply like this: -e coord=<proj>.<env>"
    when: coord is not defined

# Register k8s for state host to cache state upon completion
- hosts: state

  tasks:
  - name: "Resolve k8s for {{ inventory_hostname }}"
    k8s_client:
      server: "{{ k8s_server }}"
      namespace: "{{ k8s_namespace }}"
      username: "{{ k8s_username | default(omit) }}"
      token: "{{ k8s_token | default(omit) }}"
      distro: "{{ k8s_distro }}"
      verify_ssl: "{{ k8s_verify_ssl | default(omit) }}"
      storage_class_name: "{{ k8s_storage_class_name | default(omit) }}"
    register: k8s

# We turn our coord into group intersections
- hosts: "{{ coord.split('.') | join( ':&' ) }}"

  tasks:
  - name: "Resolve k8s for {{ inventory_hostname }}"
    k8s_client:
      server: "{{ k8s_server }}"
      namespace: "{{ k8s_namespace }}"
      username: "{{ k8s_username | default(omit) }}"
      token: "{{ k8s_token | default(omit) }}"
      distro: "{{ k8s_distro }}"
      verify_ssl: "{{ k8s_verify_ssl | default(omit) }}"
      storage_class_name: "{{ k8s_storage_class_name | default(omit) }}"
    register: k8s
    when: deployment_type == 'cluster'

  - name: "Deploy environment to {{ coord }}"
    include_role:
      name: portal_environment
```

My OS:
```text
$ cat /etc/redhat-release
CentOS Linux release 7.8.2003 (Core)
```

My `ansible`
```text
$ ansible --version
ansible 2.9.9
  config file = /home/ltheisen/git/caasd-portal-config/ansible.cfg
  configured module search path = ['/home/ltheisen/git/caasd-portal-config/cluster-ansible/lib/modules', '/home/ltheisen/git/caasd-portal-config/lib/modules']
  ansible python module location = /home/ltheisen/ansible-python/lib64/python3.6/site-packages/ansible
  executable location = /home/ltheisen/ansible-python/bin/ansible
  python version = 3.6.8 (default, Apr  2 2020, 13:34:55) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

My `ansible-lint`
```text
$ ansible-lint --version
ansible-lint 4.2.0
```

For my playbook, if i comment out the final `include_role` task it does not fail.

I installed everything in a fresh `venv` just now so everything should be up to the most recent release...

**Comment by briantist** [comment]
I'm also seeing this problem. Ansible 2.9.7, fresh venv.

**Comment by webknjaz** [comment]
Somebody needs to reproduce this with the master branch or at least a pre-release.

**Comment by briantist** [comment]
@webknjaz maybe I could try that, could you help me figure out how to run this from master?

Do you have any specific reason to believe it's been fixed or is this just ruling that out?

**Comment by webknjaz** [comment]
> @webknjaz maybe I could try that, could you help me figure out how to run this from master?

`pip install https://github.com/ansible/ansible-lint` should do. Also `pip install --pre -U ansible-lint` would install a pre-release.

> Do you have any specific reason to believe it's been fixed or is this just ruling that out?

Both, because I recall this band-aid applied a few months ago: https://github.com/ansible/ansible-lint/pull/518.

**Comment by briantist** [comment]
Playbook for testing:
```yaml
---
- hosts: localhost
  gather_facts: no
  tasks:
    - set_fact:
        one: "{{ vars | to_nice_yaml }}"
        two: "{{ vars | to_yaml }}"
```

Tried with master, `ansible-lint --version`
```
ansible-lint 4.3.0a2.dev25+geca19dd
```
✔️ no error

With pre-release:
```
ansible-lint 4.3.0a1
```
✔️ no error

---

So here's something interesting, my original version, `4.2.0`:
Python2: ✔️ no error
Python3: ❌ error

So might explain why some people see this and some don't.

It would be useful if `ansible-lint --version` also showed the pythin version like `ansible --version` does; or at least if it always uses the same version that ansible does (I have both versions and ansible seems to use 3 when both 2 and 3 are available, but it doesn't seem like `ansible-lint` is doing that).

In any case, it seems the problem (or at least the symptom) is fixed in pre-release and master versions. Thanks!

**Comment by ssbarnea** [comment]
@webknjaz please close this bug, it does not happen with the latest versions.
